### PR TITLE
Manage FluidSynth rendering and playback using the RW queue [2/4]

### LINKED ISSUE #735: Improve FluidSynth support
Continuation of #262

During #262 we identified few areas that need improvements - especially with the goal of supporting MIDI emulation on lower-specced computers like Raspberry Pi with possibly slow storage via an SD card.

Needed improvements:

1. Putting fluidsynth synthesis on a separate thread. Right now synthesis thread is shared with emulator (which is fast enough for handling synth even on older PCs, but fluidsynth complains about it and it slows down emulation). This work might require quite severe improvements to DOSBox mixer implementation, but this should pay dividends when integrating MT-32.
2. FluidSynth library has a quite extensive dependency tree via glib - they have some important reasons for doing so (e.g. POSIX is missing high-resolution timers required). On Linux it's not a big issue, but it causes problems with packaging for Windows and macOS. We should cooperate with FluidSynth team to find a way to replace those dependencies (it might be possible e.g. via C++11).
3. Slow storage mediums (slow HDDs or SD cards) when used with medium-to-large soundfonts can cause synthesis issues when emulator performs disk access - perhaps placing fluidsynth initialization in a separate thread can prevent this.
4. SoundFont loading on slows down the emulator startup (visibly - by ~0.1-0.2s) - maybe solution to (1) or (3) will solve this problem, but we need to consider improving splash screen implementation, so the soundfont loading would happen in parallel.

If anyone has spare time to tackle any of these problems: feel free to start working on them! :)

### PR DESCRIPTION
**Pull request by kcgen**
Code is exactly as we arranged for MT-32 and in a similar all-or-nothing commit (there was no way I could implement it in parts). This PR plus the background-SoundFont-loading PR will close #735.

### Background

Prior to this commit, DOSBox's interaction with FluidSynth was not threaded. FluidSynth _does use threads internally_, and it maintains a 64-byte rendering buffer sufficient to hold 1.3 ms worth audio (assuming a 48 KHz playback rate). Therefore, playback in DOSBox could  block for as much time is needed to render the requested number of frames, assuming FluidSynth's internal thread could not maintain a near-perfect realtime rate. Blocking will occur if the sysex passage requires greater-than-realtime rendering time across a period of greater than 1.3 ms.

Fortunately on modern hardware DOSBox usually has time to spare within its 1-ms tick loop, so some (sub-millisecond) blocking can be tolerated without consequence. However, rendering that takes more than a half-milliseond or more will typically generate an audio gap.

### Summary

This commit moves the FluidSynth rendering task into a thread and uses back-pressure from the RW queue to pace the rendering and consuming threads as follows:

1. Playback waits until a rendered buffer is available before passing the samples onto the mixer.

2. Rendering waits until space is available in the queue before rendering another full buffer.

To determine the total number of frames to manage in the queue, we use FluidSynth's own default audio driver settings (period sizes and period duration) to rendering up to 4096 frames in advance (~5 video frames at 60 FPS), which are divided into eight buffers of 512-frames each. To word this another way: we now have identical latency as FluidSynth's application does. (This also matches mt32emu's recommended sizes as well). 

Rendering is now performed in efficient large chunks instead of on-demand or in per-millisecond quantities. The RW queue also performs std::move() operations on the rendered buffers to ensure no copies are created.

### Testing

 - [x] Linux x86-64 HW
 - [x] Linux ARM HW
 - [x] Windows 10 (VM & HW)
 - [X] macOS  

### Performance

Two minutes of performance metrics (call-stack-sampling) were collected using this command:

 `dosbox xwin.bat & sudo perf record -p $! -g -- sleep 120`

These were converted to FlameGraphs (in SVG format) [:chart: midi-flamegraphs.zip](https://github.com/dosbox-staging/dosbox-staging/files/5975123/midi-flamegraphs.zip)

I've circles the FluidSynth call stacks.

- Before:
    

- After:
    

[Brendan Gregg explains](http://www.brendangregg.com/FlameGraphs/cpuflamegraphs.html):
> _"The x-axis spans the sample population. It does not show the passing of time from left to right, as most graphs do. The left to right ordering has no meaning (it's sorted alphabetically to maximize frame merging)."_
>
> _"The width of the box shows the total time it was on-CPU or part of an ancestry that was on-CPU (based on sample count). **Functions with wide boxes may consume more CPU per execution than those with narrow boxes, or, they may simply be called more often.** The call count is not shown (or known via sampling)."_

Last emphasis is mine. Because the width of the FluidSynth call stacks are more narrow in the PR graph, my guess is we're seeing the effect of making fewer calls into the rendering functions, given we're now requesting FluidSynth render 512 instead of ~45 (or 48) frames per-call.

### CODE REVIEW

**Comment by kcgen** [comment]
A PR to address the destruct-after-move issue is in the works, and the one-line change has been send for PR upstream: https://github.com/cameron314/readerwriterqueue/pull/113

**Comment by kcgen** [comment]
Next steps:
1. ebase this after we merge the vector PR.
2. Increase the analysis warning count by one to allow the issue flagged in the RW queue (given it provides very nice move scemantics across operations, and according to it's author the destructor-after-move prevent UB in possibly poorly written template T classes).

**Comment by kcgen** [comment]
Finished up testing. Here's a recording from macOS:
 - :movie_camera:  [rwqueue + FluidSynth (still using `std::arrays`)](https://drive.google.com/file/d/1DG47pJYhlicBGfoSjloy1psU4j2CMBrS/view?usp=sharing)

**Comment by kcgen** [comment]
Ready for review!
 - Rebased
 - Added commit justifying Clang analysis limit bump (regarding RW queue explicit destruction)
 - Retested
 
I've gone over the code several times to catch any improvements covered in the the prior PR, and can't see anything left.

File: src/midi/midi_fluidsynth.cpp
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp:60]: Why this value got removed? It's not relevant to change that is being implemented here.

I am not totally *against* this change, but it should probably be done with separate commit with explanation *why* it is being implemented.

**Review by dreamer — COMMENTED** [review]
Only a single question.

In testing I see no improvement in practice from this PR. Testing in HoMM2 seemed like it's worse now but it might be due to my old laptop or other causes (very like) - it's hard to measure.

Tested in Paku Paku, Doom, X-COM, and HoMM2.

As for flamegraphs: your analysis in unfortunately incorrect - CPU used by FluidSynth just moved to another tower (not the one you marked on image) and generated more CPU load in SDL2 code communicating with PulseAudio. In other words: based on these graphs we can't decide which solution is better.

**Comment by kcgen** [comment]
> Only a single question.
> 
> In testing I see no improvement in practice from this PR. 

Side-by-side, the same number of frames are being rendered in both solutions; so this makes a lot of sense. 

The difference between the two is that the original solution renders in 1 x 64-frame chunks (with some percent of that work done on a thread), 
while the PR renders in 8 x 512-frame chunks 100% on a thread. 

Soft limiting runs in-line with the main thread (meaning it blocks) in the original code, while soft limiting runs as part of the rendering thread in the PR.

> in HoMM2 seemed like it's worse now but it might be due to my old laptop or other causes (very like) - it's hard to measure.

Can you frequency-lock your laptop at a slow speed and/or place the dosbox primary thread along side a cpu-consuming tasks, like say OpenSSL doing crypto passes?  

Comparing the two, the PR will (or should) afford the main thread more CPU cycles. 

> 
> Tested in Paku Paku, Doom, X-COM, and HoMM2.
> 
> As for flamegraphs: your analysis in unfortunately incorrect - CPU used by FluidSynth just moved to another tower (not the one you marked on image)

Can you include screenshots to visually show this, so I can learn how to see these differences? 

> and generated more CPU load in SDL2 code communicating with PulseAudio.

There is no logical explanation for this.  The interaction at the mixer-level and downstream (into SDL) hasn't changed. The mixer still requests ~48 frames per callback, then mixes then with all the other channels, and then writes them to SDL. 

The difference is that that PR can actually be rendering the next 48 frames (and more) fully in parallel while all of that interaction with SDL is happening on the main thread. Or, if the thread ha already populate the ring, then it just sits idle. 

The original code needs to rendering the next 48 frames (64 existing - 48 frames taken = 16 frames left, leaves a new gap of 48 frames in it's single buffer) while the mixer is interacting with SDL passing it the current 1ms worth of audio.  So if anything steals CPU time, it will impinge on this next round of rendering. 

Where as the PR has a bigger slush fund of frames to operate from, and can afford to draw down of those frames during periods of limited CPU cycles. 

> In other words: based on these graphs we can't decide which solution is better.

If only looked at from a CPU usage perspective. Then I agree. I will find some scenarios that put the above theory into practice.

File: src/midi/midi_fluidsynth.cpp
Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp:60]: Right; the PR moves FluidSynth to a dedicated thread which allows FluidSynth's internal synthesizing behavior to proceed in pure linear non-threaded fashion (with respect to what FluidSynth is doing). 

This linear workflow is always the most efficient in terms of minimizing overhead; the more you split the same amount of work up by threads, the greater amount of time is spent splitting and combining the results.

That works great if we can be sure atleast one single core is fast enough to keep our ring filled with rendered buffers. 

This should be the case on 'normal and modern' hardware, but I can  imagine a case where one core is never fast enough: say a tiny SBC operating at 200 MHz per core, but has 4 cores. 

Such a tiny device might need the synthesis spread across 3 threads to keep up while DOSBox operates on the 4th core. 

Another option would be to just set this at 2 internally (and drop the setting in the conf file), knowing that splitting the synth into two threads still isn't burning up much extra overhead, and should cover 99.99% of realistic user's hardware.. even folks older Rpi2's will be easily accommodated.

Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:60]: *Sorry, I should've specified earlier. This ^^^^ is the only question I have about this PR.* :)

**Comment by dreamer** [comment]
> Can you include screenshots to visually show this, so I can learn how to see these differences?

tower-diff

Left blue line: SDL2 PulseAudio backend went from appearing in 880 (1.65%) of samples to 1520 (2.88%). No idea why, but it in the second test almost twice as much time was spent in there. Maybe you had some other audio clients in the background? Anyway, this change is not that problematic, but it means *all other towers* are thinner - hence we can't look at them and say "hey, it's thinner, therefore better!" :)

Right blue line: SoftLimiter went from being sampled in tower "on top" of "ChannelMixing" to a new tower in graph below. So ChannelMixing tower (the one you marked in your screenshot) "thinner" - not necessarily because we spent much less time in mixing, but *maybe* because we have new tower from new thread in here and consequently *other towers* got squeezed.

*When you open SVG report in browser, it has a "search field" in the top right corner - it's quite helpful.*

Maybe if we had flame graphs separated by thread… `perf` can record that, but I don't know if flamegraph scripts can.

Anyway - I don't find this a problem or a blocker for PR - just let's not use it as a justification for change. Behaviour is different now - that's all we can say based on these two flame graphs.

> > In other words: based on these graphs we can't decide which solution is better.
> 
> If only looked at from a CPU usage perspective. Then I agree. I will find some scenarios that put the above theory into practice.

Well, flamegraphs show only CPU usage and nothing more… I guess we both need to learn how to use them better :)

File: src/midi/midi_fluidsynth.cpp
Comment (dreamer) [inline src/midi/midi_fluidsynth.cpp:434]: The more I think about this solution, the more problematic it becomes…

So we have a rwqueue that can handle up to 8 "buffers". No more. And we keep churning-in - allocating, filling with data, and pushing new vectors in there again and again… moving vectors is not a problem - that happens quickly; we might be wasting resources with re-allocating and freeing vectors over and over again (when it's totally unnecessary).

*Do not change this yet. Let's merge this in - after merge we'll have at least the same solution in both MT-32 and FluidSynth. After that we'll think what's the best move forward and we both will be able to do some testing.*

Comment (kcgen) [reply src/midi/midi_fluidsynth.cpp:434]: You're description was brilliant here @dreamer " we keep churning-in - allocating, filling with data, and pushing new vectors in there again and again… moving vectors is not a problem - that happens quickly; we might be wasting resources with re-allocating and freeing vectors over and over again (when it's totally unnecessary)."

Fully agree - I think there's one last optimization step that will nail this:

1) We know the rendering phase into the vector<float>'s already good, because it only uses a single vector for the entire lifetime.
2) The re-allocation happens in the SoftLimiter, where it creates the vector<int16_t> temporary, which is eventually moved into the queue, moved out of the queue, and then eventually destroyed.

If we passed a `& vector<int16_t>` into the SoftLimiter for it to write into, then we could keep recycling it on the outside.  

> Do not change this yet. Let's merge this in - after merge we'll have at least the same solution in both MT-32 and FluidSynth. 

Absolutely :-)  We've got a working baseline, and you're headed to take it to the next level!

**Comment by kcgen** [comment]
Thanks for the help with interpreting the flame-graphs! 

I recall there were text breakdowns that can be generated against the folded perf data - that might help in this case we were just want to be sure we're on par (or maybe better).

What the graphs can't show is how little time was spent blocking DOSBox's main thread inside the 1ms loop.  We're trying to drive that blocking time down to zero, especially during periods of typical background CPU usage - ie: user has their web-browser opened with a bunch of tabs, and some pesky javascript fires up every couple second and spikes Firefox's CPU usage.

Those 8 buffers' worth sitting in the ring means the PR code can be "starved" for a 64-fold longer duration than the baseline  (4096 frames vs. 64 frames).  Is this ability to ride out marginal CPU periods and eliminate blocking on the main-thread where benefits reside.

File: src/midi/midi_fluidsynth.cpp
Comment (dreamer) [reply src/midi/midi_fluidsynth.cpp:60]: @kcgen Hey, I am quite sure your comment did not exist last time I responded in this thread… Maybe some problem on GitHub side, whatever :)

> This linear workflow is always the most efficient in terms of minimizing overhead; (…)

Not always! **This PR** removes linear workflow to minimize overhead in main thread! Having FluidSynth do synthesis in a single thread *might be* appropriate, but *we don't know that yet* (!).

> the more you split the same amount of work up by threads, the greater amount of time is spent splitting and combining the results.

That's called algorithm scalability; *We don't know* how scalable internal FluidSynth algorithm really is. It might work ok, it might work not-so-ok, it might depend on user's hardware, it might depend on soundfont being used… 

Maybe we should indeed avoid exposing this setting to the end-users, but *we already do expose it*. I don't want to remove this setting just to bring it back in 0.78.0 if it'll turn out to be necessary for someone.

Can we bring this setting back, please? Maybe we'll remove it in the future (likely by replacing with some heuristic that will pick good value for user *somehow*), but I think we need it for now.

**Comment by kcgen** [comment]
### EDIT IN PROGRESS (saving)

Star Wars X Wing Collector's CD-ROM

 - Configured with Sound Blaster SFX
 - MIDI Music



[cpu]
core=dynamic
cycles=10000

[dosbox]
machine=svga_s3
memsize=4

[gus]
gus=false

[dos]
xms=true
ems=true
umb=false

[speaker]
pcspeaker=false
tandy=off
disney=false

[midi]
mididevice = fluidsynth
[mixer]
rate = 48000

[fluidsynth]
soundfont = ../sf2/AudioFabric-01797.sf2



``` python
import sys
import statistics as s

y = list()
for line in sys.stdin:
  if '' == line.strip():
    break
  y.append(float(line.strip()[:-1]))

mean = s.mean(y)
median = s.median(y)
stdev = s.stdev(y)
print("%8.2f %8.2f %8.2f | %8.2f" % (median, mean,  stdev, max(y)))
```

``` shell
grep ^[0-9].*, log.txt | python3 stats.py 
```

PR branch

| test | Synth-threads | median | average | std-deviation | max | log-file |
| --- | --- | --- | --- | --- | --- | --- |
| Idle | 1 | 1.81 |    3.06 | 8.91 | 1367.52 |  [pr-idle.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011828/pr-idle.txt)  |  
| Idle | 2 | 1.83 |  3.05 |  5.26 | 393.72 | [pr-idle-2t.txt](https://github.com/dosbox-staging/dosbox-staging/files/6012345/pr-idle-2t.txt) |
| 1x Load | 1 | 3.01 |    5.37  |   44.42 | 5136.05 | [pr-single-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011829/pr-single-load.txt) | 
| 1x Load | 2 | 3.04  |   5.95 |  56.09 | 4998.57 | [pr-single-load-2t.txt](https://github.com/dosbox-staging/dosbox-staging/files/6012368/pr-single-load-2t.txt) |
| 1x Load | 3 | 3.07  |  5.47  |  39.28 | 2896.65 | [pr-single-load-3t.txt](https://github.com/dosbox-staging/dosbox-staging/files/6012369/pr-single-load-3t.txt) |
| 2x Load | 1 | 4.25  |  37.34 | 1905.69 | 224870.29 | [pr-double-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011830/pr-double-load.txt)  |
| 2x Load | 2 |4.44 |  117.40 | 6562.72 | 616023.29 | [pr-double-load-2t.txt](https://github.com/dosbox-staging/dosbox-staging/files/6012381/pr-double-load-2t.txt) |
| 2x Load | 3 |4.41 |  98.81 | 4884.97 | 522232.93 | [pr-double-load-3t.txt](https://github.com/dosbox-staging/dosbox-staging/files/6012382/pr-double-load-3t.txt) |

Master branch

| test | median | average | std-deviation | max | log-file |
| --- | --- | --- | --- | --- | --- |
| Idle |  108.36 |  143.76 |  143.96 | 4710.85 | [master-idle.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011838/master-idle.txt) | 


| 1x Load | 171.70 |  267.75 |  693.57 | 39241.54 | [master-single-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011839/master-single-load.txt) |
| 2x Load |  90.51  | 509.99 | 8130.03 | 473482.28 | [master-double-load.txt](https://github.com/dosbox-staging/dosbox-staging/files/6011840/master-double-load.txt) |

**Review by dreamer — APPROVED** [review]
Thanks, @kcgen! I think that's enough evidence :)

**Pull request merged**
